File: engine/schemeMatcher.ts
/**
 * Deterministic scheme compatibility matcher — Phase 2 P1.
 *
 * Pure functions. No AI. No network. No EMI/DSCR calculations.
 * Evaluates scheme-rule compatibility based on profile data.
 * Must produce identical results to backend/app/services/scheme_matcher.py.
 */

import { Decimal } from 'decimal.js';

const TWO_PLACES = 2;

// ── Types ───────────────────────────────────────────────────────────────────

export interface MatchProfile {
  project_cost: number | null;
  requested_loan_amount: number | null;
  location_type: string | null;
  business_category: string | null;
}

export interface SchemeRuleInput {
  id: string;
  scheme_id: string;
  rule_version: number;
  min_project_cost: number;
  max_project_cost: number;
  max_loan_amount: number;
  active: boolean;
  effective_from: string | null; // ISO date YYYY-MM-DD
  effective_to: string | null;
  location_type: string | null;
  allowed_business_categories: string[] | string | null; // JSON array or parsed
  last_verified_at: string | null; // ISO datetime
}

export interface FailedRule {
  field: string;
  reason: string;
  actual?: number | string;
  allowed_min?: number;
  allowed_max?: number;
  required?: string;
  allowed?: string[];
}

export interface FreshnessPolicy {
  stale_after_hours: number;
}

export interface MatchResult {
  status: 'COMPATIBLE' | 'NOT_COMPATIBLE' | 'INSUFFICIENT_DATA';
  scheme_id: string;
  scheme_rule_id: string;
  rule_version: number;
  validity: string;
  freshness: string;
  passed_rules: string[];
  failed_rules: FailedRule[];
  missing_fields: string[];
  warnings: string[];
  evaluated_at: string;
  selection_reason?: string;
  specificity?: number;
  competing_rules_count?: number;
}

// ── Validity ────────────────────────────────────────────────────────────────

export function evaluateRuleValidity(
  rule: SchemeRuleInput,
  evaluationDate: string,
): string {
  if (!rule.active) return 'INACTIVE';

  if (rule.effective_from) {
    if (rule.effective_from > evaluationDate) return 'NOT_YET_EFFECTIVE';
  }
  if (rule.effective_to) {
    if (rule.effective_to < evaluationDate) return 'EXPIRED';
  }

  return 'ACTIVE';
}

// ── Freshness ───────────────────────────────────────────────────────────────

export function evaluateRuleFreshness(
  rule: SchemeRuleInput,
  evaluationDt: Date,
  policy: FreshnessPolicy,
): string {
  if (!rule.last_verified_at) return 'UNKNOWN';

  const lastVerified = new Date(rule.last_verified_at);
  if (isNaN(lastVerified.getTime())) return 'UNKNOWN';

  const ageHours = (evaluationDt.getTime() - lastVerified.getTime()) / (1000 * 3600);
  return ageHours <= policy.stale_after_hours ? 'VERIFIED_FRESH' : 'VERIFIED_STALE';
}

// ── Specificity ─────────────────────────────────────────────────────────────

const OPTIONAL_CONSTRAINT_FIELDS = ['location_type', 'allowed_business_categories'] as const;

export function ruleSpecificity(rule: SchemeRuleInput): number {
  let count = 0;
  if (rule.location_type != null) count++;
  if (rule.allowed_business_categories != null) count++;
  return count;
}

// ── Single-rule evaluation ──────────────────────────────────────────────────

export function evaluateSchemeCompatibility(
  profile: MatchProfile,
  schemeRule: SchemeRuleInput,
  evaluationDate: string, // YYYY-MM-DD
  freshnessPolicy: FreshnessPolicy,
): MatchResult {
  const nowUtc = new Date();
  const validity = evaluateRuleValidity(schemeRule, evaluationDate);
  const freshness = evaluateRuleFreshness(schemeRule, nowUtc, freshnessPolicy);

  const passed: string[] = [];
  const failed: FailedRule[] = [];
  const missing: string[] = [];
  const warnings: string[] = [];

  // ── Validity constraints ──
  if (validity === 'INACTIVE') {
    failed.push({ field: 'active', reason: 'RULE_INACTIVE' });
  } else if (validity === 'NOT_YET_EFFECTIVE') {
    failed.push({ field: 'effective_from', reason: 'NOT_YET_EFFECTIVE' });
  } else if (validity === 'EXPIRED') {
    failed.push({ field: 'effective_to', reason: 'EXPIRED' });
  } else {
    passed.push('active_status');
    passed.push('effective_dates');

    // ── Project cost ──
    if (profile.project_cost == null) {
      missing.push('project_cost');
    } else {
      const pc = new Decimal(profile.project_cost).toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);
      const minPc = new Decimal(schemeRule.min_project_cost).toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);
      const maxPc = new Decimal(schemeRule.max_project_cost).toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);

      if (pc.lt(minPc)) {
        failed.push({
          field: 'project_cost', reason: 'BELOW_MINIMUM',
          actual: pc.toNumber(), allowed_min: minPc.toNumber(),
        });
      } else if (pc.gt(maxPc)) {
        failed.push({
          field: 'project_cost', reason: 'ABOVE_MAXIMUM',
          actual: pc.toNumber(), allowed_max: maxPc.toNumber(),
        });
      } else {
        passed.push('project_cost_range');
      }
    }

    // ── Requested loan amount ──
    if (profile.requested_loan_amount == null) {
      missing.push('requested_loan_amount');
    } else {
      const rl = new Decimal(profile.requested_loan_amount).toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);
      const maxLoan = new Decimal(schemeRule.max_loan_amount).toDecimalPlaces(TWO_PLACES, Decimal.ROUND_HALF_UP);
      if (rl.gt(maxLoan)) {
        failed.push({
          field: 'requested_loan_amount', reason: 'ABOVE_MAXIMUM',
          actual: rl.toNumber(), allowed_max: maxLoan.toNumber(),
        });
      } else {
        passed.push('loan_amount');
      }
    }

    // ── Location type (optional constraint) ──
    if (schemeRule.location_type != null) {
      if (profile.location_type == null) {
        missing.push('location_type');
      } else if (profile.location_type.toUpperCase() !== schemeRule.location_type.toUpperCase()) {
        failed.push({
          field: 'location_type', reason: 'LOCATION_MISMATCH',
          actual: profile.location_type, required: schemeRule.location_type,
        });
      } else {
        passed.push('location_type');
      }
    }

    // ── Business category (optional constraint) ──
    const ruleCatsRaw = schemeRule.allowed_business_categories;
    if (ruleCatsRaw != null) {
      const ruleCats: string[] = typeof ruleCatsRaw === 'string'
        ? JSON.parse(ruleCatsRaw)
        : ruleCatsRaw;
      const ruleCatsUpper = ruleCats.map(c => c.toUpperCase());

      if (profile.business_category == null) {
        missing.push('business_category');
      } else if (!ruleCatsUpper.includes(profile.business_category.toUpperCase())) {
        failed.push({
          field: 'business_category', reason: 'CATEGORY_NOT_ALLOWED',
          actual: profile.business_category, allowed: ruleCats,
        });
      } else {
        passed.push('business_category');
      }
    }
  }

  // ── Freshness warnings ──
  if (freshness === 'VERIFIED_STALE') warnings.push('USING_STALE_SCHEME_DATA');
  else if (freshness === 'UNKNOWN') warnings.push('UNKNOWN_FRESHNESS');

  // ── Final status (precedence: FAIL > MISSING > PASS) ──
  let status: 'COMPATIBLE' | 'NOT_COMPATIBLE' | 'INSUFFICIENT_DATA';
  if (failed.length > 0) {
    status = 'NOT_COMPATIBLE';
  } else if (missing.length > 0) {
    status = 'INSUFFICIENT_DATA';
  } else {
    status = 'COMPATIBLE';
  }

  return {
    status,
    scheme_id: schemeRule.scheme_id,
    scheme_rule_id: schemeRule.id,
    rule_version: schemeRule.rule_version,
    validity,
    freshness,
    passed_rules: passed,
    failed_rules: failed,
    missing_fields: missing,
    warnings,
    evaluated_at: nowUtc.toISOString(),
  };
}

// ── Multi-rule selection ────────────────────────────────────────────────────

function effectiveFromOrdinal(rule: SchemeRuleInput): number {
  if (!rule.effective_from) return 0;
  return new Date(rule.effective_from + 'T00:00:00Z').getTime();
}

function ruleSortKey(rule: SchemeRuleInput): [number, number, number, string] {
  return [
    -ruleSpecificity(rule),
    -effectiveFromOrdinal(rule),
    -(rule.rule_version || 0),
    rule.id,
  ];
}

function compareKeys(a: [number, number, number, string], b: [number, number, number, string]): number {
  for (let i = 0; i < 3; i++) {
    if (a[i] !== b[i]) return (a[i] as number) - (b[i] as number);
  }
  return a[3] < b[3] ? -1 : a[3] > b[3] ? 1 : 0;
}

export function selectBestRule(
  compatibleResults: MatchResult[],
  rulesById: Record<string, SchemeRuleInput>,
): MatchResult | null {
  if (compatibleResults.length === 0) return null;

  if (compatibleResults.length === 1) {
    const result = { ...compatibleResults[0] };
    result.selection_reason = 'ONLY_COMPATIBLE_RULE';
    result.specificity = ruleSpecificity(rulesById[result.scheme_rule_id]);
    result.competing_rules_count = 1;
    return result;
  }

  const sorted = [...compatibleResults].sort((a, b) => {
    const ka = ruleSortKey(rulesById[a.scheme_rule_id]);
    const kb = ruleSortKey(rulesById[b.scheme_rule_id]);
    return compareKeys(ka, kb);
  });

  const winner = { ...sorted[0] };
  const winnerRule = rulesById[winner.scheme_rule_id];
  const runnerUpRule = rulesById[sorted[1].scheme_rule_id];

  const wSpec = ruleSpecificity(winnerRule);
  const rSpec = ruleSpecificity(runnerUpRule);

  let reason: string;
  if (wSpec > rSpec) {
    reason = 'HIGHEST_SPECIFICITY';
  } else {
    const wEff = effectiveFromOrdinal(winnerRule);
    const rEff = effectiveFromOrdinal(runnerUpRule);
    if (wEff > rEff) {
      reason = 'NEWEST_EFFECTIVE_FROM';
    } else if ((winnerRule.rule_version || 0) > (runnerUpRule.rule_version || 0)) {
      reason = 'HIGHEST_RULE_VERSION';
    } else {
      reason = 'UUID_TIE_BREAK';
    }
  }

  winner.selection_reason = reason;
  winner.specificity = wSpec;
  winner.competing_rules_count = compatibleResults.length;
  return winner;
}

// ── Multi-scheme evaluation ─────────────────────────────────────────────────

export interface SchemeGroupResult {
  compatible_schemes: Array<{
    scheme_id: string;
    best_rule: MatchResult | null;
    all_compatible_rules_count: number;
  }>;
  incompatible_schemes: Array<{
    scheme_id: string;
    best_result: MatchResult;
  }>;
  insufficient_data_schemes: Array<{
    scheme_id: string;
    best_result: MatchResult;
  }>;
  status?: string;
}

export function evaluateAllSchemes(
  profile: MatchProfile,
  rules: SchemeRuleInput[],
  evaluationDate: string,
  freshnessPolicy: FreshnessPolicy,
): SchemeGroupResult {
  if (!rules || rules.length === 0) {
    return {
      compatible_schemes: [],
      incompatible_schemes: [],
      insufficient_data_schemes: [],
      status: 'INSUFFICIENT_SCHEME_DATA',
    };
  }

  const resultsByScheme: Record<string, MatchResult[]> = {};
  const rulesById: Record<string, SchemeRuleInput> = {};

  for (const rule of rules) {
    rulesById[rule.id] = rule;
    const result = evaluateSchemeCompatibility(profile, rule, evaluationDate, freshnessPolicy);
    const sid = rule.scheme_id;
    if (!resultsByScheme[sid]) resultsByScheme[sid] = [];
    resultsByScheme[sid].push(result);
  }

  const compatible_schemes: SchemeGroupResult['compatible_schemes'] = [];
  const incompatible_schemes: SchemeGroupResult['incompatible_schemes'] = [];
  const insufficient_data_schemes: SchemeGroupResult['insufficient_data_schemes'] = [];

  for (const [schemeId, results] of Object.entries(resultsByScheme)) {
    const compatible = results.filter(r => r.status === 'COMPATIBLE');
    const insufficient = results.filter(r => r.status === 'INSUFFICIENT_DATA');

    if (compatible.length > 0) {
      const best = selectBestRule(compatible, rulesById);
      compatible_schemes.push({
        scheme_id: schemeId,
        best_rule: best,
        all_compatible_rules_count: compatible.length,
      });
    } else if (insufficient.length > 0) {
      insufficient_data_schemes.push({
        scheme_id: schemeId,
        best_result: insufficient[0],
      });
    } else {
      incompatible_schemes.push({
        scheme_id: schemeId,
        best_result: results[0],
      });
    }
  }

  return { compatible_schemes, incompatible_schemes, insufficient_data_schemes };
}

File: backend/app/services/scheme_matcher.py
"""
Deterministic scheme compatibility matcher — Phase 2 P1.

Pure functions. No AI. No DB. No network. No EMI/DSCR calculations.
Evaluates scheme-rule compatibility based on profile data.
"""
import json
from datetime import date, datetime, timezone
from decimal import Decimal, ROUND_HALF_UP
from typing import Any, Dict, List, Optional, Tuple

TWO_PLACES = Decimal("0.01")

# ── Validity ─────────────────────────────────────────────────────────────────

def evaluate_rule_validity(
    rule: Dict[str, Any],
    evaluation_date: date,
) -> str:
    """Return ACTIVE | NOT_YET_EFFECTIVE | EXPIRED | INACTIVE."""
    if not rule.get("active", True):
        return "INACTIVE"

    eff_from = rule.get("effective_from")
    eff_to = rule.get("effective_to")

    if eff_from:
        if isinstance(eff_from, str):
            eff_from = date.fromisoformat(eff_from)
        if eff_from > evaluation_date:
            return "NOT_YET_EFFECTIVE"

    if eff_to:
        if isinstance(eff_to, str):
            eff_to = date.fromisoformat(eff_to)
        if eff_to < evaluation_date:
            return "EXPIRED"

    return "ACTIVE"


# ── Freshness ────────────────────────────────────────────────────────────────

def evaluate_rule_freshness(
    rule: Dict[str, Any],
    evaluation_dt: datetime,
    freshness_policy: Dict[str, Any],
) -> str:
    """Return VERIFIED_FRESH | VERIFIED_STALE | UNKNOWN."""
    last_verified = rule.get("last_verified_at")
    if last_verified is None:
        return "UNKNOWN"

    if isinstance(last_verified, str):
        last_verified = datetime.fromisoformat(last_verified.replace("Z", "+00:00"))

    if not isinstance(last_verified, datetime):
        return "UNKNOWN"

    stale_hours = freshness_policy.get("stale_after_hours", 720)
    age_hours = (evaluation_dt - last_verified).total_seconds() / 3600.0
    return "VERIFIED_FRESH" if age_hours <= stale_hours else "VERIFIED_STALE"


# ── Specificity ──────────────────────────────────────────────────────────────

OPTIONAL_CONSTRAINT_FIELDS = ("location_type", "allowed_business_categories")

def rule_specificity(rule: Dict[str, Any]) -> int:
    """Count non-null optional constraint fields (not metadata)."""
    return sum(1 for f in OPTIONAL_CONSTRAINT_FIELDS if rule.get(f) is not None)


# ── Single-rule evaluation ───────────────────────────────────────────────────

def evaluate_scheme_compatibility(
    profile: Dict[str, Any],
    scheme_rule: Dict[str, Any],
    evaluation_date: str | date,
    freshness_policy: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Evaluate a single scheme rule against a profile.

    Returns {status, validity, freshness, passed_rules, failed_rules,
             missing_fields, warnings, scheme_id, scheme_rule_id, rule_version,
             evaluated_at}.
    """
    if isinstance(evaluation_date, str):
        evaluation_date = date.fromisoformat(evaluation_date)

    now_utc = datetime.now(timezone.utc)

    validity = evaluate_rule_validity(scheme_rule, evaluation_date)
    freshness = evaluate_rule_freshness(scheme_rule, now_utc, freshness_policy)

    passed: List[str] = []
    failed: List[Dict[str, Any]] = []
    missing: List[str] = []
    warnings: List[str] = []

    # ── Validity constraints (short-circuit) ──
    if validity == "INACTIVE":
        failed.append({"field": "active", "reason": "RULE_INACTIVE"})
    elif validity == "NOT_YET_EFFECTIVE":
        failed.append({"field": "effective_from", "reason": "NOT_YET_EFFECTIVE"})
    elif validity == "EXPIRED":
        failed.append({"field": "effective_to", "reason": "EXPIRED"})
    else:
        passed.append("active_status")
        passed.append("effective_dates")

        # ── Project cost ──
        project_cost_raw = profile.get("project_cost")
        if project_cost_raw is None:
            missing.append("project_cost")
        else:
            pc = Decimal(str(project_cost_raw)).quantize(TWO_PLACES, ROUND_HALF_UP)
            min_pc = Decimal(str(scheme_rule["min_project_cost"])).quantize(TWO_PLACES, ROUND_HALF_UP)
            max_pc = Decimal(str(scheme_rule["max_project_cost"])).quantize(TWO_PLACES, ROUND_HALF_UP)

            if pc < min_pc:
                failed.append({
                    "field": "project_cost", "reason": "BELOW_MINIMUM",
                    "actual": float(pc), "allowed_min": float(min_pc),
                })
            elif pc > max_pc:
                failed.append({
                    "field": "project_cost", "reason": "ABOVE_MAXIMUM",
                    "actual": float(pc), "allowed_max": float(max_pc),
                })
            else:
                passed.append("project_cost_range")

        # ── Requested loan amount ──
        req_loan_raw = profile.get("requested_loan_amount")
        if req_loan_raw is None:
            missing.append("requested_loan_amount")
        else:
            rl = Decimal(str(req_loan_raw)).quantize(TWO_PLACES, ROUND_HALF_UP)
            max_loan = Decimal(str(scheme_rule["max_loan_amount"])).quantize(TWO_PLACES, ROUND_HALF_UP)
            if rl > max_loan:
                failed.append({
                    "field": "requested_loan_amount", "reason": "ABOVE_MAXIMUM",
                    "actual": float(rl), "allowed_max": float(max_loan),
                })
            else:
                passed.append("loan_amount")

        # ── Location type (optional constraint) ──
        rule_loc = scheme_rule.get("location_type")
        if rule_loc is not None:
            profile_loc = profile.get("location_type")
            if profile_loc is None:
                missing.append("location_type")
            elif profile_loc.upper() != rule_loc.upper():
                failed.append({
                    "field": "location_type", "reason": "LOCATION_MISMATCH",
                    "actual": profile_loc, "required": rule_loc,
                })
            else:
                passed.append("location_type")

        # ── Business category (optional constraint) ──
        rule_cats_raw = scheme_rule.get("allowed_business_categories")
        if rule_cats_raw is not None:
            rule_cats = rule_cats_raw if isinstance(rule_cats_raw, list) else json.loads(rule_cats_raw)
            rule_cats_upper = [c.upper() for c in rule_cats]
            profile_cat = profile.get("business_category")
            if profile_cat is None:
                missing.append("business_category")
            elif profile_cat.upper() not in rule_cats_upper:
                failed.append({
                    "field": "business_category", "reason": "CATEGORY_NOT_ALLOWED",
                    "actual": profile_cat, "allowed": rule_cats,
                })
            else:
                passed.append("business_category")

    # ── Freshness warnings ──
    if freshness == "VERIFIED_STALE":
        warnings.append("USING_STALE_SCHEME_DATA")
    elif freshness == "UNKNOWN":
        warnings.append("UNKNOWN_FRESHNESS")

    # ── Final status (precedence: FAIL > MISSING > PASS) ──
    if failed:
        status = "NOT_COMPATIBLE"
    elif missing:
        status = "INSUFFICIENT_DATA"
    else:
        status = "COMPATIBLE"

    return {
        "status": status,
        "scheme_id": scheme_rule.get("scheme_id"),
        "scheme_rule_id": scheme_rule.get("id"),
        "rule_version": scheme_rule.get("rule_version"),
        "validity": validity,
        "freshness": freshness,
        "passed_rules": passed,
        "failed_rules": failed,
        "missing_fields": missing,
        "warnings": warnings,
        "evaluated_at": now_utc.isoformat(),
    }


# ── Multi-rule selection ─────────────────────────────────────────────────────

def _rule_sort_key(rule: Dict[str, Any]) -> Tuple:
    """Sort key for deterministic rule selection (best first)."""
    spec = rule_specificity(rule)
    eff_from = rule.get("effective_from")
    if eff_from:
        if isinstance(eff_from, str):
            eff_from = date.fromisoformat(eff_from).toordinal()
        elif isinstance(eff_from, date):
            eff_from = eff_from.toordinal()
        else:
            eff_from = 0
    else:
        eff_from = 0

    version = rule.get("rule_version", 0)
    rule_id = rule.get("id", "")

    return (-spec, -eff_from, -version, rule_id)


def select_best_rule(
    compatible_results: List[Dict[str, Any]],
    rules_by_id: Dict[str, Dict[str, Any]],
) -> Optional[Dict[str, Any]]:
    """Select one representative rule from a list of COMPATIBLE evaluation results."""
    if not compatible_results:
        return None

    if len(compatible_results) == 1:
        result = compatible_results[0]
        result["selection_reason"] = "ONLY_COMPATIBLE_RULE"
        result["specificity"] = rule_specificity(rules_by_id[result["scheme_rule_id"]])
        result["competing_rules_count"] = 1
        return result

    # Sort using the deterministic key
    sorted_results = sorted(
        compatible_results,
        key=lambda r: _rule_sort_key(rules_by_id[r["scheme_rule_id"]]),
    )

    winner = sorted_results[0]
    winner_rule = rules_by_id[winner["scheme_rule_id"]]
    runner_up = sorted_results[1]
    runner_up_rule = rules_by_id[runner_up["scheme_rule_id"]]

    # Determine selection reason
    w_spec = rule_specificity(winner_rule)
    r_spec = rule_specificity(runner_up_rule)
    if w_spec > r_spec:
        reason = "HIGHEST_SPECIFICITY"
    else:
        w_eff = winner_rule.get("effective_from")
        r_eff = runner_up_rule.get("effective_from")
        w_eff_ord = date.fromisoformat(w_eff).toordinal() if isinstance(w_eff, str) and w_eff else 0
        r_eff_ord = date.fromisoformat(r_eff).toordinal() if isinstance(r_eff, str) and r_eff else 0
        if w_eff_ord > r_eff_ord:
            reason = "NEWEST_EFFECTIVE_FROM"
        elif winner_rule.get("rule_version", 0) > runner_up_rule.get("rule_version", 0):
            reason = "HIGHEST_RULE_VERSION"
        else:
            reason = "UUID_TIE_BREAK"

    winner["selection_reason"] = reason
    winner["specificity"] = w_spec
    winner["competing_rules_count"] = len(compatible_results)
    return winner


# ── Multi-scheme evaluation ──────────────────────────────────────────────────

def evaluate_all_schemes(
    profile: Dict[str, Any],
    rules: List[Dict[str, Any]],
    evaluation_date: str | date,
    freshness_policy: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Evaluate profile against all scheme rules, group by scheme.
    Returns {compatible_schemes, incompatible_schemes, insufficient_data_schemes}.
    Does NOT rank schemes against each other.
    """
    if not rules:
        return {
            "compatible_schemes": [],
            "incompatible_schemes": [],
            "insufficient_data_schemes": [],
            "status": "INSUFFICIENT_SCHEME_DATA",
        }

    # Evaluate each rule
    results_by_scheme: Dict[str, List[Dict[str, Any]]] = {}
    rules_by_id: Dict[str, Dict[str, Any]] = {}

    for rule in rules:
        rules_by_id[rule["id"]] = rule
        result = evaluate_scheme_compatibility(profile, rule, evaluation_date, freshness_policy)
        sid = rule.get("scheme_id", "")
        if sid not in results_by_scheme:
            results_by_scheme[sid] = []
        results_by_scheme[sid].append(result)

    compatible_schemes = []
    incompatible_schemes = []
    insufficient_data_schemes = []

    for scheme_id, results in results_by_scheme.items():
        compatible = [r for r in results if r["status"] == "COMPATIBLE"]
        insufficient = [r for r in results if r["status"] == "INSUFFICIENT_DATA"]

        if compatible:
            best = select_best_rule(compatible, rules_by_id)
            compatible_schemes.append({
                "scheme_id": scheme_id,
                "best_rule": best,
                "all_compatible_rules_count": len(compatible),
            })
        elif insufficient:
            insufficient_data_schemes.append({
                "scheme_id": scheme_id,
                "best_result": insufficient[0],
            })
        else:
            incompatible_schemes.append({
                "scheme_id": scheme_id,
                "best_result": results[0],
            })

    return {
        "compatible_schemes": compatible_schemes,
        "incompatible_schemes": incompatible_schemes,
        "insufficient_data_schemes": insufficient_data_schemes,
    }
